New mentoring notes: python armstrong-numbers (#2110)
exercism/website-copy: tracks/python/exercises/armstrong-numbers/mentoring.md

# Mentoring

## Problem and challenges

> The problem requires the student to sum each digit of a number raised to the (length of number) power.

Reviewer: ```suggestion
The problem requires the student to sum each digit of an integer or _natural number_ raised to the number of digits' power.
```

The length of number may be clear enough.

## Reasonable solutions

> This exercise can be solved fairly straight forward:

Reviewer 2: ```suggestion
This exercise can be solved in a fairly straightforward fashion:
```

```python
def is_armstrong(number):
    digits = (int(d) for d in str(number))
    count = len(str(number))
    total = sum(digit ** count for digit in digits)
    return number == total
```

> This may also be written without the generator.

Reviewer 2: ```suggestion
This solution can be written without the generator.
```

> This avoid the more complex generator type at the expense of a more complex the sum line.

Reviewer: ```suggestion
This avoids the more complex generator type at the expense of a more complex sum line.
```

```python
def is_armstrong(number):
    count = len(str(number))
    total = sum(int(digit) ** count for digit in str(number))
    return number == total
```

### Common Suggestions

#### Generator Expressions

> ```python
>     total = sum([digit ** count for digit in digits])
> ```

Reviewer 2: ```suggestion
```python
    # With an intermediate list passed to sum.
    total = sum([digit ** count for digit in digits])
    
    # With a generator expression passed to sum
    total = sum(digit ** count for digit in digits)
```
```

> Often students are unaware of [generator expressions](https://www.python.org/dev/peps/pep-0289/) and that they can be passed directly to functions that accept iterables.

Reviewer 2: ```suggestion
Students are often unaware of [generator expressions](https://www.python.org/dev/peps/pep-0289/) and that they can be passed directly to functions that accept iterables.
```

> This uses a generator expression to build a list then iterates through the list.

Reviewer 2: ```suggestion
The first line uses a comprehension to build a `list`  for `sum()` to iterate through.
```

> The middle step of a temporarily list is not needed.

Reviewer: ```suggestion
The middle step of a temporary list is not needed.
```

> The middle step of a temporary list is not needed.

Reviewer 2: ```suggestion
The middle step of a temporary list takes up space in memory, and is unneeded.
The second line shows the `list` replaced by an equivalent `generator expression` that `sum()` can also iterate through. 
```

Potential feedback snippet:

```
`[f(a) for a in b]` iterates through `b` to create a list. You are then passing this list to a function which iterates through that list to build a result and discards that list. You can avoid that extra middle step of building and storing a list by passing a generator to the function instead and having the function process the elements of that iterator directly! You can do this either by doing `func(f(a) for a in b)` or `generator = (f(a) for a in b); func(generator)`

> ```

Reviewer: Consider avoiding "you" for "one", or "the code".

Such as "This passe this list to a function which iterates through that list to build a result and discard that list.  This avoids an extra middle step of building…"

It avoids unnecessary assignment of personality or ego and tends to focus more on the code for the reader, the voice being less _accusatory_.

Author: Changed.

Reviewer: Was it changed somewhere else?

### Talking points

> Students unfamiliar with `sum()` may use a for loop and sum up the values themselves:

Reviewer: ```suggestion
Students unfamiliar with `sum()` may use a for loop to add the values themselves:
```

```python
def is_armstrong(number):
    count = len(str(number))
    total = 0
    for digit in str(number):
        total += int(digit) ** count
    return number == total
```

> Suggest the students have a look at Python's [built-in functions](https://docs.python.org/3/library/functions.html) for a helper function that can make this easier.

Reviewer: ```suggestion
Suggest having a look at Python's [built-in functions](https://docs.python.org/3/library/functions.html) for a helper function that can make this easier.
```

# Mentoring

## Problem and challenges

The problem requires the student to sum each digit of an integer or _natural number_ raised to the number of digits' power.

## Reasonable solutions

This exercise can be solved in a fairly straightforward fashion:

```python
def is_armstrong(number):
    digits = (int(d) for d in str(number))
    count = len(str(number))
    total = sum(digit ** count for digit in digits)
    return number == total
```

This solution can also be written without the generator.
This avoids the more complex generator type at the expense of a more complex sum line.

```python
def is_armstrong(number):
    count = len(str(number))
    total = sum(int(digit) ** count for digit in str(number))
    return number == total
```

### Common Suggestions

#### Generator Expressions

```python
    # With an intermediate list passed to sum.
    total = sum([digit ** count for digit in digits])
    
    # With a generator expression passed to sum
    total = sum(digit ** count for digit in digits)

> Students are often unaware of [generator expressions](https://www.python.org/dev/peps/pep-0289/) and that they can be passed directly to functions that accept iterables.

Reviewer: If you want to do so now, reference links are preferred according to the Markdown Guide for Exercism.

Author: I'd prefer to defer that change and let it happen with the mass fix PR.

Reviewer: I was unaware of a mass fix PR that would happen.  I will defer to that as well then.

The first line uses a comprehension to build a `list`  for `sum()` to iterate through.
The middle step of a temporary list takes up space in memory and is not needed.
The second line shows the `list` replaced by an equivalent `generator expression` that `sum()` can also iterate through. 

Potential feedback snippet:

```
`[f(a) for a in b]` iterates through `b` to create a list. You are then passing this list to a function which iterates through that list to build a result and discards that list. You can avoid that extra middle step of building and storing a list by passing a generator to the function instead and having the function process the elements of that iterator directly! You can do this either by doing `func(f(a) for a in b)` or `generator = (f(a) for a in b); func(generator)`
```

### Talking points

Students unfamiliar with `sum()` may use a for loop to add the values themselves:

```python
def is_armstrong(number):
    count = len(str(number))
    total = 0
    for digit in str(number):
        total += int(digit) ** count
    return number == total
```

> Suggest having a look at Python's [built-in functions](https://docs.python.org/3/library/functions.html) for a helper function that can make this easier.

Reviewer: Reference link here as well.

Author: Deferring
